### ISSUE TITLE
Tracking issue for DST coercions (coerce_unsized, unsize) stabilization

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#27732]
[DST coercions](https://github.com/rust-lang/rfcs/pull/982) were the last part of DST to land, and are still undergoing [some changes](https://github.com/rust-lang/rfcs/pull/1234).

Before stabilization, at the very least the API surface here needs some scrutiny from the libs team for naming and other conventions issues.

cc @nrc @Gankro 

## Things to consider prior to stabilization

* Interaction with pin, see [this internals thread](https://internals.rust-lang.org/t/unsoundness-in-pin/11311) and #68015 for details.

**Comment by aturon** [comment]
@nrc, what's your sense of the status for these features? Do you feel ready to stabilize after the libs team takes a look?

**Comment by nrc** [comment]
My feeling (and it is really just a feeling, I have no data to back this up) is that DST coercions haven't yet got the use they need to prove themselves for stabilisation. I'd love to be wrong about this - if someone could show me a bunch of code doing stuff with DST coercions I'd be very happy. OTOH, I don't think there is anything fundamentally wrong with them, so it would not be a waste of time for the libs team to take a look.

**Comment by photino** [comment]
Why is the marker named `Unsize` but not `Unsized`?

**Comment by arielb1** [comment]
@photino 

It goes the wrong way around - for example, we have `[i8; 2]: Unsize<[u8]>` (and also `[i8; 2]: Unsize<fmt::Debug>`). We *can* swap the order of parameters - `U: Unsized<T>` if `U` is the unsized version of `T`.

**Comment by rphmeier** [comment]
Can we move to stabilize this? It's been unchanged for 5 months and I've heard no complaints about it.

**Comment by Gankro** [comment]
I don't recall having any issues with this in libstd. The correct way to use this is basically a witchcraft incantation (mostly because these are under-the-covers things one never interacts with in real code), but as far as I know, there's know way to use it *dangerously* wrong? The whole thing Just Works.

I guess one thing I've never really thought about is why one needs to talk about whether the pointed-to value is unsizeable or whatever. Why can't one simply declare that pointer X is an unsizable and have all the junk automatic? Is this desirable to have for generic consumers...?

**Comment by jnicholls** [comment]
It's desirable for generic consumers of e.g. Arena allocators that allocate a Sized type that implements a trait unknown to the allocator, and the consumer wants to coerce the allocated result to an allocated trait object.

See https://gitlab.com/Bastacyclop/rust_arena/blob/master/src/lib.rs in the tests.  It's a decent use case.  How would you recommend doing this otherwise?

**Comment by stevenblenkinsop** [comment]
Is the lack of support for enums deliberate, temporary, or merely overlooked?

**Comment by aturon** [comment]
I'm nominating this for discussion at the next lang team meeting.

**Comment by Gankro** [comment]
@jnicholls maybe I'm missing something, but just mark the one pointer field that will be Unsized as coercable: https://gitlab.com/Bastacyclop/rust_arena/blob/master/src/lib.rs#L41

It seems to me that all the `impl<...> CoerceUnsized for ...` is trivially derivable from the claim that that field should be coercable.

**Comment by jnicholls** [comment]
@Gankro Sorry I don't follow what you mean.  Can you provide an example?

**Comment by Gankro** [comment]
Oops, something clobbered my selection, I meant to link to: https://gitlab.com/Bastacyclop/rust_arena/blob/master/src/lib.rs#L59

Basically instead of 

```rust
pub struct Allocated<'a, 'b, T: 'b + ?Sized> {
    reference: &'b mut T,
    phantom: PhantomData<&'a ()>
}

impl<'a, 'b, T: ?Sized, U: ?Sized> CoerceUnsized<Allocated<'a, 'b, U>> for Allocated<'a, 'b, T>
    where U: 'b, T: 'b + Unsize<U> {}
```

You would just do *something* like:

```
#[coerce_unsized(T)]
pub struct Allocated<'a, 'b, T: 'b + ?Sized> {
    reference: &'b mut T,
    phantom: PhantomData<&'a ()>
}
```

**Comment by aturon** [comment]
cc @eddyb

**Comment by aturon** [comment]
Removed T-libs tag, but note that we'd like to review the naming conventions here before stabilization.

**Comment by eddyb** [comment]
There's room for generalization and a good chance we might want `Source: Coerce<Target>`.
See [this comment](https://github.com/rust-lang/rfcs/issues/1403#issuecomment-166980781) in response to wanting `&mut T` reborrows working through structs.

**Comment by apasel422** [comment]
Right now, because `T` does not implement `Unsize<T>`, one cannot use `CoerceUnsized` to support optional coercions in a generic context. (My use case is supporting DSTs with the placement API). Since `&T` can trivially be coerced to `&T`, would it make sense to add the identity impl?

**Comment by tomaka** [comment]
I'm not familiar with how this works, but shouldn't the `T` of `CoerceUnsized` be `T: ?Sized`?
Right now `T` is required to be `Sized`, which doesn't make sense for me. [Example](http://is.gd/Sj2tdT).

**Comment by arielb1** [comment]
@tomaka 

You implement `CoerceUnsized` on pointer types, not on DST types.

**Comment by alexcrichton** [comment]
(adding T-libs as this has a library API associated with it which needs stabilization)

**Comment by nikomatsakis** [comment]
So at a recent @rust-lang/lang meeting we discussed this. I think everyone felt...somewhat comfortable with the idea of going forward, but we also wanted to do a thorough review of the current status of the code, along with a comparison against the existing specifications, to make sure everything is well aligned.

**Comment by eddyb** [comment]
@nikomatsakis The problem, though, is that the RFC is *intentionally provisional*, mostly to get `Rc` and `Arc` working with DSTs. I would like to see an attempt at generalization before stabilizing it (I believe the compiler code would "just work" with more than one field being coerced, and various coercion modes, not just `Unsize`, with minimal changes).

Maybe with specialization we can even encode `Deref` coercions in a general `Coerce` trait.

**Comment by arielb1** [comment]
@eddyb 

`Deref` coercions involve transitivity, which the trait system does not try to support.

**Comment by abonander** [comment]
@nrc triage ping

**Comment by nrc** [comment]
No change. I think we are waiting for some kind of custom DST RFC before moving forward with stabilisation here.

**Comment by H2CO3** [comment]
@nrc @Gankro Any progress with this? It's really annoying that my beloved custom smart pointers can't point to trait objects in the lack of stable `Unsize` and `CoerceUnsized`.

**Comment by omni-viral** [comment]
What instead of stabilizing `Unsize` and `CoerceUnsized` we could implement `From` for all pointers from `std` which implement `CoerceUnsized`.
For example.
```
impl<T, U> From<Box<T>> for Box<U>
where
  T: Unsize<U>
{
  fn from(value: Box<T>) -> Box<U> {
    value as Box<U>
  }
}
```
In this case we could implement `From` for our pointer wrappers relying on `From` implementation of pointers.
And there will be no issues with converting multiple fields of the struct.

**Comment by Osspial** [comment]
It's been over a year since anyone's really looked at this. What's stopping this from getting stabilized?

**Comment by eddyb** [comment]
The current design is perma-unstable. A new RFC is required for a design that we can consider stabilizing.

The `Unsize` trait can be ignored, as `T: Unsize<U>` and `*mut T: CoerceUnsized<*mut U>` are pretty much equivalent.

So what we should do is generalize `CoerceUnsized` (rename to `Coerce` and handle more/all kinds of coercions through it).

Then we can allow even `impl<T: Coerce<U>, U> Coerce<Option<U>> for Option<T>`, which has been requested for a long while, too.

IMO "custom DSTs" was a red herring, as stabilizing `Unsize` is unnecessary for making coercions work.

**Comment by eddyb** [comment]
Nominating for discussion, regarding splitting this issue into:
* `Unsize` trait, reflects unsizing, may want to wait on "custom DSTs", irrelevant for "opt into unsizing coercions", can stay unstable forever (or even removed)
* `CoerceUnsized`, can be generalized to an "opt into structural coercion" `Coerce` trait, would cover all the cases "`CoerceUnsized` with `Unsize` bounds" cover today (and more), can be stabilized more readily

**Comment by Mcat12** [comment]
I just wanted to note (if it hasn't been noted already) that this feature makes it possible to write a struct with a generic trait parameter. This was useful for me because I could then require the other generics to implement the generic trait.

Related internals thread I found helpful:
https://internals.rust-lang.org/t/traits-as-generic-type-parameters/4950

Example (a mockable Rocket request guard):
```rust
pub enum Service<'r, Trait, Impl, Mock>
where
    Trait: ?Sized + 'r,
    Impl: Unsize<Trait> + FromRequest<'a, 'r>,
    Mock: Deref + FromRequest<'a, 'r>,
    <Mock as Deref>::Target: Unsize<Trait>
{
    Prod(Impl, PhantomData<&'r Trait>),
    Test(Mock)
}

pub type MyServiceGuard<'r> = Service<
    'r,
    MyServiceTrait,
    MyServiceImpl<'r>,
    rocket::State<'r, MyServiceMock>
>;
```

**Comment by mathstuf** [comment]
This feature is how I'm currently performing `arc_of_t as arc_of_u` given `T: U` (the type needs extra traits in the application, but a library doesn't know about that trait and just requires a "base" trait). Is there some other way of doing this on stable?

**Comment by eddyb** [comment]
Came across this again and... how did https://github.com/rust-lang/rust/issues/27732#issuecomment-480075569 result in nothing? Not even a comment?

Re-nominating (low priority, but should still not be unnominated before a decision).

**Comment by nikomatsakis** [comment]
@eddyb we discussed in @rust-lang/lang meeting today -- splitting off into two feature gates makes sense, provided we have a good description of what the two gates cover and the interaction between them. (e.g., as the issue comments)

**Comment by matthieu-m** [comment]
Has any consideration been given to having `T: Unsize<T>` for `Sized` types?

When attempting to use `Unsize` in generic contexts, [such as in this example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c9fb1acc121d41ca157d8172160ad13a), the inability to abstract over sized and unsized types requires splitting APIs in 2, with the same code in each, just because `T: !Unsize<T>`.

**Comment by Chronos-Sk** [comment]
@matthieu-m That playground example is currently throwing an ICE, in case that's new. [Error gist](https://gist.github.com/Chronos-Sk/c335c49fc20724efed8623b18db85c3f).

**Comment by matthieu-m** [comment]
> @matthieu-m That playground example is currently throwing an ICE, in case that's new. [Error gist](https://gist.github.com/Chronos-Sk/c335c49fc20724efed8623b18db85c3f).

I _think_ this is new, but as I mentioned because `T: !Unsize<T>` it didn't compile in the first place.

**Comment by eggyal** [comment]
Currently, there's no std/core type for a "sized str" (by which I mean an array of UTF-8 bytes).  Of course, one can create a custom version:

```rust
use core::str;

pub struct Str<const N: usize>([u8; N]);

impl<const N: usize> Str<N> {
    pub fn new(s: [u8; N]) -> Result<Self, str::Utf8Error> {
        str::from_utf8(&s)?;
        Ok(Self(s))
    }

    pub unsafe fn new_unchecked(s: [u8; N]) -> Self {
        Self(s)
    }
}
```

But such a custom type cannot be "unsized" to `str`.  Shouldn't there be a library type like this, but that implements `Unsize<str>`?

**Comment by bjorn3** [comment]
```rust
impl<const N: usize> Deref for Str<N> {
    type Target = str;
    
    fn deref(&self) -> &str {
        unsafe {
            str::from_utf8_unchecked(&self.0)
        }
    }
}
```

allows you to "unsize" `&Str<N>` to `&str` using `&*s`.

**Comment by eggyal** [comment]
@bjorn3 Of course, but that's not a coercion! In particular, given `s: Str<N>`, one cannot write `&s as &str`—and hence when coercing generic `T: Unsize<U>` with `&t as &U` we cannot provide a `Str<N>` for `t`.

**Comment by JanBeh** [comment]
I have a [problem where I want to be generic over reference-like types](https://users.rust-lang.org/t/generalized-references-and-unsized-coercion/76056), and I feel like the current mechanism doesn't allow me something what I would like to do.

**Comment by RReverser** [comment]
> one cannot write `&s as &str`

Yes, they can, that's what Deref provides. Your example of `&s as &str` compiles just fine: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bd56db69728b59c755a066c9e42ef37c

**Comment by eggyal** [comment]
> Yes, they can, that's what Deref provides.

Perhaps my previous comment wasn't clear enough: a deref coercion is not an unsize coercion.  What you cannot do is call the following with a `&Str<N>` argument where `U` is `str`:

```rust
fn unsize<U>(t: &impl Unsize<U>) -> &U {
    t
}
```

**Comment by dest1n1s** [comment]
Is there something wrong with the rule

> Types implementing a trait `Trait` also implement `Unsize<dyn Trait>`.

when it works with a trait object with associate type, such as `dyn Trait<Assoc=Type>`?

For example:

```rust
trait Foo {
    fn foo(&self);
}

trait Bar: Foo {
    type Assoc;

    fn bar(&self);
}

fn upcast<Dyn>(bar: &Dyn) -> &dyn Foo
where
    Dyn: ?Sized + Unsize<dyn Foo>,
{
    bar
}

impl Foo for i32 {
    fn foo(&self) {
        println!("foo");
    }
}

impl Bar for i32 {
    type Assoc = i32;

    fn bar(&self) {
        println!("bar");
    }
}

fn main() {
    let bar: &dyn Bar<Assoc = i32> = &42;
    let foo = upcast(bar);
    foo.foo();
}
```

 in which the compiler says 

> the trait `Unsize<(dyn Foo + 'static)>` is not implemented for `dyn Bar<Assoc = i32>`

But if I try to manually impl the trait `Foo` for `dyn Bar<Assoc=i32>`, it says

> `(dyn Bar<Assoc = i32> + 'static)` automatically implements trait `Foo`

Here's the minimal example: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=85925cd4977fc95beb52ce9a03653755 ,

**Comment by Chronos-Sk** [comment]
FWIW, that's effectively relying on the incomplete feature `trait_upcasting` (which seems to have the same difficulty: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=fb3a77efd3979ea2d94b5177ded51750)).

**Comment by nikomatsakis** [comment]
Closing in favor of https://github.com/rust-lang/rust/issues/18598 -- they seem to cover the same ground.
